src/util/timelock.rs

```diff
@@ -0,0 +1,264 @@
+// Rust Bitcoin Library
+// Written in 2022 by
+//     Tobin C. Harding <me@tobin.cc>
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Bitcoin timelocks.
+//!
+//! Relative and absolute timelocks based on `nSequence` and `nLockTime`.
+//!
+
+#[cfg(feature = "std")]
+use std::error;
+
+use core::fmt;
+
+use crate::blockdata::constants::{
+    LOCKTIME_THRESHOLD, SEQUENCE_LOCKTIME_DISABLE_FLAG, SEQUENCE_LOCKTIME_GRANULARITY,
+    SEQUENCE_LOCKTIME_TYPE_FLAG,
+};
+
+/// An absolute timelock, expires after either block height or time.
+///
+/// This is nLockTime as defined in BIP-65.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Absolute {
+    /// A block height absolute timelock.
+    Block {
+        height: u32,
+    },
+    /// A UNIX timestamp absolute timelock.
+    Time {
+        seconds: u32,
```

Comment:
This allows creating inconsistent values: `let evil = Absolute::Block { height: LOCKTIME_THRESHOLD, }`
I'd like to have more newtypes:

```rust
// enforce < LOCKTIME_THRESHOLD in constructors
pub struct Height(u32);

// enforce values,
// also can contain conversions to seconds and such
pub struct Time(u32);
```

Reply:
Enforcing correct values is an additional argument for introducing these types (at the cost of well... having more types to worry about). With these types `Transaction::get_maximum_relative_timelocks` can return `(Option<Height>, Option<Time>)` avoiding need to document it and eliminating possibility of mistake.

`fn into_relative()` can be used to make turning them back into timelock types easier.